[FLINK-5041] Savepoint backwards compatibility 1.1 -> 1.2

StefanRRichter opened the pull request:
This PR implements sub-issues [FLINK-5042] and [FLINK-5043] of [FLINK-5041].

[FLINK-5042] : Backwards compatibility for savepoints is implemented by reintroducing some older classes from 1.1 that have been removed/replaced/changed in 1.2. All such classes are moved to a different packet prefixed as `org.apache.flink.migration`, while their class names remain unchanged. With `MigrationInstantiationUtil` a special deserialization facility is provided that allows to load Java serialized objects of classes from 1.1. The idea is to redirect their fully qualified classname to their new location in `org.apache.flink.migration`. This is done through a modified `ObjectInputStream`. All state meta structures and handles are then converted to 1.2.

[FLINK-5043] : Introduced code in the keyed state backends ensures backwards compatibility by translating states from the old backends based on Heap, Filesystem, and RocksDB.

StefanRRichter commented:
R please @aljoscha @StephanEwen or who else is interested in taking a look into this.

StephanEwen commented on flink-core/src/main/java/org/apache/flink/core/memory/ByteArrayInputStreamWithPos.java:
```diff
@@ -0,0 +1,268 @@
+import java.io.InputStream;
+
+/**
+ * Un-synchronized copy of Java's ByteArrayOutputStream that also exposes the current position.
```
I think we cannot copy this code, as it comes from a source with incompatible license.
I think you need to re-create this class without relying on the source.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/migration/runtime/checkpoint/KeyGroupState.java:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	public void discard(ClassLoader classLoader) throws Exception {
+		keyGroupState.deserializeValue(classLoader).discardState();
```
Let's "deactivate" this method by throwing an `UnsupportedOperationException`.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/migration/runtime/checkpoint/SubtaskState.java:
```diff
@@ -0,0 +1,108 @@
+	}
+
+	public void discard(ClassLoader userClassLoader) throws Exception {
+		state.deserializeValue(userClassLoader).discardState();
```
Deactivate this method (see above)

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/migration/runtime/checkpoint/TaskState.java:
```diff
@@ -0,0 +1,166 @@
+	}
+
+	public void discard(ClassLoader classLoader) throws Exception {
+		for (SubtaskState subtaskState : subtaskStates.values()) {
```
Deactivate this method (see above)

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/migration/runtime/state/AbstractStateBackend.java:
```diff
@@ -0,0 +1,68 @@
+
+		@Override
+		public void discardState() throws Exception {
+			stream.discardState();
```
Deactivate this method (see above)

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/migration/state/MigrationKeyGroupStateHandle.java:
```diff
@@ -0,0 +1,39 @@
+
+@Internal
+@Deprecated
+public class MigrationKeyGroupStateHandle extends KeyGroupsStateHandle {
```
Quick comment that this class' purpose is to tag a handle.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/migration/streaming/runtime/tasks/StreamTaskState.java:
```diff
@@ -0,0 +1,177 @@
+	 *                   state handles and snapshots.
+	 */
+	public void discardState() throws Exception {
+		StateHandle<?> operatorState = this.operatorState;
```
Deactivate this method (see above)

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/migration/streaming/runtime/tasks/StreamTaskStateList.java:
```diff
@@ -0,0 +1,121 @@
+
+	@Override
+	public void discardState() throws Exception {
+		for (StreamTaskState state : states) {
```
Deactivate this method (see above)

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapKeyedStateBackend.java:
```diff
@@ -346,4 +359,130 @@ public String toString() {
+			return;
+		}
+
+		HashMap<String, KvStateSnapshot<K, ?, ?, ?>> namedStates =
```
Sanity check that the collection contains no more than one element.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapKeyedStateBackend.java:
```diff
@@ -346,4 +359,130 @@ public String toString() {
+				namespaceSerializer = VoidNamespaceSerializer.INSTANCE;
+			}
+
+			Map nullNameSpaceFix = (Map) rawResultMap.remove(null);
```
This looks like it should be guarded by a test ;-)

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapKeyedStateBackend.java:
```diff
@@ -346,4 +359,130 @@ public String toString() {
+
+			DataInputViewStreamWrapper inView = new DataInputViewStreamWrapper(inStream);
+
+			final int numKeys = inView.readInt();
```
Minor comment: It seems this variable actually means `numNamespaces`.

StephanEwen commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1317,9 +1320,13 @@ class JobManager(
+                    jobId,
+                    executionGraph.getAllVertices,
+                    savepointPath,
+                    userCodeLoader,
```
Minor comment: I think you can grab the user code ClassLoader form the ExecutionGraph (that allows you to keep the class loader variable final and in the narrower scope)

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/migration/runtime/checkpoint/savepoint/SavepointV0.java:
```diff
@@ -0,0 +1,102 @@
+
+	@Override
+	public void dispose() throws Exception {
+		ClassLoader cl = Thread.currentThread().getContextClassLoader();
```
This is only used to dispose the Savepoint in our tests.

StefanRRichter commented:
Thanks a lot for the review, I addressed the comments (except for some tests which will still follow in a different commit). Looking for approval on the changes/rebase and somebody to merge this. Could you help me on this @uce ?

aljoscha commented:
@StefanRRichter ist there already an issue for adding a very elaborate test for this that restores from a Flink 1.1 savepoint? If not, could you please create one.

Very good work, by the way! 👍 😃

StefanRRichter commented:
Thanks @aljoscha ! I will talk with @twalthr about this today and will create a JIRA after we agreed upon the scope and way of testing this. Expect a JIRA later today.
